icu4c/source/test/cintltst/ucptrietest.c

```diff
@@ -705,6 +707,59 @@ testBuilder(const char *testName, const UMutableCPTrie *mutableTrie,
     testTrieGetRanges(testName, NULL, mutableTrie, UCPMAP_RANGE_NORMAL, 0, checkRanges, countCheckRanges);
 }
 
+static void
+trieTestGolden(const char *testName,
+             const UCPTrie* trie,
+             const CheckRange checkRanges[],
+             int32_t countCheckRanges) {
+    log_verbose("golden testing Trie '%s'\n", testName);
+
+    UErrorCode status = U_ZERO_ERROR;
+    const char *testdatapath = loadSourceTestData(&status);
+    char goldendatapath[512];
+    // note: snprintf always writes a NUL terminator.
+    snprintf(goldendatapath, 512, "%sucptrie%s%s.toml",
+        testdatapath, U_FILE_SEP_STRING, testName);
+
+    // Write the data into a tmpfile (memstream is not portable)
+    FILE* stream = tmpfile();
+    fputs("[ucptrie.struct]\n", stream);
+    usrc_writeUCPTrie(stream, testName, trie, UPRV_TARGET_SYNTAX_TOML);
+    fputs("\n[ucptrie.testdata]\n", stream);
+    fputs("# Array of (limit, value) pairs\n", stream);
+    usrc_writeArray(stream, "checkRanges = [\n\t",
+        // Note: CheckRange is a tuple of two 32-bit words
+        checkRanges, 32, countCheckRanges*2,
+        "\t", "\n]\n");
+
+    // Convert the stream into a memory buffer
+    long fsize = ftell(stream);
+    void* memoryBuffer = malloc(fsize + 1);
+    if (!memoryBuffer) {
+        status = U_MEMORY_ALLOCATION_ERROR;
+        log_err_status(status, "Could not allocate buffer: %s", goldendatapath);
+        goto cleanup;
+    }
+    fseek(stream, 0, SEEK_SET);
+    fread(memoryBuffer, 1, fsize, stream);
+
+    int32_t testResult = uprv_compareGoldenFiles(
+        memoryBuffer, fsize,
+        goldendatapath,
+        getTestOption(WRITE_GOLDEN_DATA_OPTION));
```

Comment:
I see, since golden data is created / overwritten in a test mode anyways, we want to create the new golden data first before failing a/the test. Is that correct?

Reply:
We create the golden data so that we can either (1) compare it with what's on the filesystem or (2) dump it to the filesystem, depending on `getTestOption(WRITE_GOLDEN_DATA_OPTION)`.